sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala

```diff
@@ -62,6 +62,9 @@ object Canonicalize extends {
     case a: Add => orderCommutative(a, { case Add(l, r) => Seq(l, r) }).reduce(Add)
     case m: Multiply => orderCommutative(m, { case Multiply(l, r) => Seq(l, r) }).reduce(Multiply)
 
+    case o: Or => orderCommutative(o, { case Or(l, r) => Seq(l, r) }).reduce(Or)
```

Comment:
Both`OR` and `AND` short circuit evaluation. This means that if you use a stateful (non-deterministic) expression in an `OR` or an `AND` you will change the result by reordering, for example: `RAND() > 0.5 OR x = 1`. The optimizer should not really change the output of a logical plan (and we have enforced this in the past).

I think we should only reorder expressions if they are deterministic.